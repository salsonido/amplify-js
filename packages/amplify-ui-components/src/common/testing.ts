import {
	boolean as booleanKnob,
	select as selectKnob,
	text as textKnob,
} from '@storybook/addon-knobs';

export const pixelThreshold = 10;

export const selectors = {
	// amplify-form-field
	formFieldDescription: '[data-test=form-field-description]',
};

export const knobs = {
	labelKnob(labelPlaceholder: string) {
		return textKnob('Label', labelPlaceholder);
	},

	descriptionKnob(descriptionPlaceholder: string) {
		return textKnob('Description', descriptionPlaceholder);
	},

	hintKnob(hintPlaceholder: string) {
		return textKnob('Hint', hintPlaceholder);
	},

	inputTypeKnob(inputTypePlaceholder: string) {
		return selectKnob(
			'Type',
			[
				'date',
				'email',
				'number',
				'password',
				'search',
				'tel',
				'text',
				'url',
				'time',
			],
			inputTypePlaceholder
		);
	},

	placeholderKnob(placeholderPlaceholder: string) {
		return textKnob('Placeholder', placeholderPlaceholder);
	},

	buttonTypeKnob(buttonTypePlaceholder: string) {
		return selectKnob(
			'Button type',
			['button', 'submit', 'reset'],
			buttonTypePlaceholder
		);
	},

	buttonTextKnob(buttonTextPlaceholder: string) {
		return textKnob('Button text', buttonTextPlaceholder);
	},

	toggleKnob(toggleDefault: boolean) {
		return booleanKnob('Toggle', toggleDefault);
	},

	disabledKnob(disabledDefault: boolean) {
		return booleanKnob('Disabled', disabledDefault);
	},

	formSectionHeaderKnob(headerPlaceholder: string) {
		return textKnob('Header text', headerPlaceholder);
	},

	formSectionSubmitKnob(submitPlaceholder: string) {
		return textKnob('Submit button text', submitPlaceholder);
	},

	tooltipAutoShowKnob(autoShowDefault: boolean) {
		return booleanKnob('Tooltip autoshow', autoShowDefault);
	},
};
