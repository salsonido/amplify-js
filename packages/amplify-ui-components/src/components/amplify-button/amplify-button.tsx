import { Element, Component, Prop, h, Host } from '@stencil/core';
import { ButtonTypes, ButtonVariant } from '../../common/types/ui-types';
import { IconNameType } from '../amplify-icon/icons';

/**
 * @slot (default) - content placed within the button
 */
@Component({
	tag: 'amplify-button',
	styleUrl: 'amplify-button.scss',
})
export class AmplifyButton {
	@Element() el!: HTMLAmplifyButtonElement;
	/** Type of the button: 'button', 'submit' or 'reset' */
	@Prop() type: ButtonTypes = 'button';
	/** Variant of a button: 'button' | 'anchor | 'icon' */
	@Prop() variant: ButtonVariant = 'button';
	/** (Optional) Callback called when a user clicks on the button */
	@Prop() handleButtonClick: (evt: Event) => void;
	/** Disabled state of the button */
	@Prop() disabled?: boolean = false;
	/** Name of icon to be placed inside the button */
	@Prop() icon?: IconNameType;

	render() {
		return (
			<Host>
				<button
					class={{
						[this.variant]: true,
					}}
					type={this.type}
					disabled={this.disabled}
					onClick={this.handleButtonClick}
				>
					{this.variant === 'icon' && (
						<amplify-icon name={this.icon}></amplify-icon>
					)}
					<slot />
				</button>
			</Host>
		);
	}
}
