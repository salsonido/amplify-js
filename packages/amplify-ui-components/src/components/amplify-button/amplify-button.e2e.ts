import { E2EPage, newE2EPage } from '@stencil/core/testing';
import { pixelThreshold } from '../../common/testing';

describe('amplify-button e2e:', () => {
	let page: E2EPage;

	beforeEach(async () => {
		page = await newE2EPage();
		await page.setContent('<amplify-button>FOO</amplify-button>');
	});

	it(`renders with text 'FOO'`, async () => {
		const element = await page.find('amplify-button');
		expect(element).not.toBeNull();

		const screenshot = await page.compareScreenshot('amplify-form-section', {
			fullPage: true,
		});
		expect(screenshot).toMatchScreenshot({
			allowableMismatchedPixels: pixelThreshold,
		});
	});

	it('fires its onButtonClick callback upon being clicked', async () => {
		const buttonElement = await page.find('amplify-button');
		expect(buttonElement).not.toBeNull();

		const func = jest.fn();
		await page.exposeFunction('exposedfunc', func);

		// This block here, and the .exposeFunction() above, are both necessary
		// if you ever want to pass a function into an object's props
		await page.$eval('amplify-button', (componentElement: any) => {
			componentElement.handleButtonClick = this.exposedfunc;
		});
		await page.waitForChanges();

		const button = await buttonElement.find('button');
		expect(func).not.toHaveBeenCalled();
		await button.click();
		expect(func).toHaveBeenCalledTimes(1);
	});
});
