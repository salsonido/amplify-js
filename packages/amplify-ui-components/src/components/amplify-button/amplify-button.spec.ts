import { newSpecPage } from '@stencil/core/testing';
import { AmplifyButton } from './amplify-button';

describe('amplify-button spec:', () => {
	describe('Render logic ->', () => {
		it('renders with no button text', async () => {
			const page = await newSpecPage({
				components: [AmplifyButton],
				html: `<amplify-button></amplify-button>`,
			});

			expect(page.root).toMatchSnapshot();
		});
		it('renders with button text FOO', async () => {
			const page = await newSpecPage({
				components: [AmplifyButton],
				html: `<amplify-button>FOO</amplify-button>`,
			});

			expect(page.root).toMatchSnapshot();
		});
		it('renders with danger class when button type is reset', async () => {
			const page = await newSpecPage({
				components: [AmplifyButton],
				html: `<amplify-button type='reset'>FOO</amplify-button>`,
			});

			expect(page.root).toMatchSnapshot();
		});
	});
	describe('Component logic ->', () => {
		it(`should have type 'button' by default`, async () => {
			const button = new AmplifyButton();
			expect(button.type).toBe('button');
		});
	});
});
