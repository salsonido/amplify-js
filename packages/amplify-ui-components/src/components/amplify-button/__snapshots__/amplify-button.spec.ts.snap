// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-button spec: Render logic -> renders with button text FOO 1`] = `
<amplify-button>
  <!---->
  <button class="button" type="button">
    FOO
  </button>
</amplify-button>
`;

exports[`amplify-button spec: Render logic -> renders with danger class when button type is reset 1`] = `
<amplify-button type="reset">
  <!---->
  <button class="button" type="reset">
    FOO
  </button>
</amplify-button>
`;

exports[`amplify-button spec: Render logic -> renders with no button text 1`] = `
<amplify-button>
  <!---->
  <button class="button" type="button"></button>
</amplify-button>
`;
